### PR TITLE
Format the output XML to match GitLab/Jenkins expectations

### PR DESCRIPTION
**Pull request by sasa-tomic**
These changes depend on updates in a dependent crate `junit-report`, which are in a pending PR.
In the meantime, I have the changes in a forked repo, but wanted to submit a PR here as well for increased visibility of the issue.

Namely, the problem is that `cargo2junit` at the moment generates XML files which do not get parsed properly by GitLab and Jenkins if the "message" attribute is too large. This happens because we're not following the specification of the JUnit XML.

### CODE REVIEW

**Comment by johnterickson** [comment]
Thanks for this - looking forward to when `junit-report` is updated and released!

**Comment by johnterickson** [comment]
> Just to clarify that I've introduced an additional limit in the test `stdout` data size to 64KB by default, since GitLab was still failing to parse some XML files in which some tests had an excessive amount of logs.
> 
> The limit of 64KB can be changed by setting the `TEST_STDOUT_MAX_LEN` environment variable to the desired value.

Thanks! Can you create a const for that value? Then tests can reference it too 😃

**Comment by darxriggs** [comment]
@sasa-tomic #34 is also an issue on Jenkins. Is this covered here? If no, maybe you are interested in fixing it too.

**Comment by johnterickson** [comment]
@sasa-tomic Is this ready for me to review for merging?

**Comment by johnterickson** [comment]
Ah looks like `junit-report` is still pointing to a GitHub repo and not a crate.  Keep up the good work - looking forward to this!

**Comment by sasa-tomic** [comment]
I'ved asked about the release here: https://github.com/bachp/junit-report-rs/issues/14

File: src/main.rs [hunk 1]
```diff
@@ -38,7 +40,11 @@ enum TestEvent {
     #[serde(rename = "ok")]
     Ok { name: String },
     #[serde(rename = "failed")]
-    Failed { name: String, stdout: String },
+    Failed {
+        name: String,
+        stdout: Option<String>,
+        stderr: Option<String>,
```
Comment (jstarks) [inline src/main.rs:46]: I don't think Rust includes `stderr` separately from `stdout`.

**Comment by johnterickson** [comment]
@sasa-tomic Can you update to reference the newly released `junit-report`? That will unblock https://github.com/johnterickson/cargo2junit/pull/41

**Comment by johnterickson** [comment]
###

**Comment by johnterickson** [comment]
Apologies- literally clicked the wrong button - I did not intend to close! :)

File: src/main.rs (hunk 1 above)
Comment (sasa-tomic) [reply src/main.rs:46]: I agree @jstarks, libtest in `cargo test` does not send a separate stderr, but we use a different test framework which provides both `stdout` and `stderr` and we tried to use cargo2junit for it as well.
In the end this proved to be much more work than desired, due to transitive dependencies to `junit-report`, and in the end we gave up and just re-implemented direct junit xml dumping from scratch.
If you think this hurts the project, please feel free to remove the lines.

**Comment by sasa-tomic** [comment]
@johnterickson should be clean now, I've resolved the conflict and updated the junit-report dependency. Sorry for the delay.

File: src/main.rs [hunk 2]
```diff
@@ -196,12 +201,60 @@ fn parse<T: BufRead>(
...
-                            TestCase::failure(&name, duration, "cargo test", &stdout)
-                                .set_classname(module_path.as_str()),
+                            TestCase::failure(
+                                &name,
+                                duration,
+                                "cargo test",
+                                &format!("failed {}::{}", module_path.as_str(), &name),
+                            )
+                            .set_classname(module_path.as_str())
+                            .set_system_out(&system_out),
```
Comment (johnterickson) [inline src/main.rs:257]: @sasa-tomic Is there a reason that both `stdout` and `stderr` are piped into `set_system_out`?  It seems that routing each to `set_system_out` and `set_system_err` respectively is a better match.

Comment (sasa-tomic) [reply src/main.rs:257]: Yes, you're right but junit-report does not provide system_err. And we didn't have the time to extend the junit-report.

Comment (sasa-tomic) [reply src/main.rs:257]: Or hmm looks like there is system_err in junit-report. I was wrong. Either I didn't look well earlier or this was added in the meantime.

Comment (sasa-tomic) [reply src/main.rs:257]: Would you mind taking over this pr and improving it? I'm quite swamped with other things atm and won't have time to work on this in the next week or two.

Comment (johnterickson) [reply src/main.rs:257]: No problem. Can you make sure that I have permission to push changes to this pr? ￼ https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

Comment (sasa-tomic) [reply src/main.rs:257]: I believe this is set already, @johnterickson  

image

Comment (svenstaro) [reply src/main.rs:257]: Friendly ping. :) I'd love to see this merged!

**Comment by johnterickson** [comment]
Will finish this over here https://github.com/johnterickson/cargo2junit/pull/43

**Pull request closed without merging**